File: 2021/07/21/AddBinary.java
/**
 * https://leetcode.com/problems/add-binary/
 * Tough 5000m race on the track.  I'm pretty slow right now.
 * @author Andrew Jarombek
 * @since 7/21/2021
 */

class AddBinary {
    public static String addBinary(String a, String b) {
        int maxLength = Math.max(a.length(), b.length());

        StringBuilder result = new StringBuilder();
        int carryOver = 0;

        for (int i = 0; i < maxLength; i++) {
            int aIndex = a.length() - 1 - i;
            int bIndex = b.length() - 1 - i;

            char aValue = i >= a.length() ? '0' : a.charAt(aIndex);
            char bValue = i >= b.length() ? '0' : b.charAt(bIndex);

            int aNumber = Character.getNumericValue(aValue);
            int bNumber = Character.getNumericValue(bValue);

            int totalNumber = aNumber + bNumber + carryOver;
            if (totalNumber == 3) {
                result.insert(0, "1");
                carryOver = 1;
            } else if (totalNumber == 2) {
                result.insert(0, "0");
                carryOver = 1;
            } else {
                result.insert(0, String.valueOf(totalNumber));
                carryOver = 0;
            }
        }

        if (carryOver == 1) {
            result.insert(0, "1");
        }

        return result.toString();
    }

    public static void main(String[] args) {
        assert AddBinary.addBinary("1", "11").equals("100");
    }
}

File: 2022/04/08/cpp/addBinary.cpp
// https://leetcode.com/problems/add-binary/
// Author: Andrew Jarombek
// Date: 4/8/2022

#include <cassert>
#include <string>
#include <stack>

using namespace std;

string addBinary(string a, string b) {
    stack<string> digits;
    int remainder = 0;

    int i = a.length() - 1;
    int j = b.length() - 1;

    while (i >= 0 || j >= 0) {
        int x = i >= 0 ? a[i] - '0' : 0;
        int y = j >= 0 ? b[j] - '0' : 0;
        int val = x + y + remainder;

        remainder = val > 1 ? 1 : 0;
        digits.push(val % 2 == 0 ? "0" : "1");

        i--;
        j--;
    }

    if (remainder > 0) {
        digits.push("1");
    }

    string result;
    while (digits.size() > 0) {
        result += digits.top();
        digits.pop();
    }

    return result;
}

int main() {
    assert(addBinary("1010", "1011") == "10101");
}